### PR TITLE
Migrate to Structured Logs in pkg/controller

### LINKED ISSUE #120: Use structured logs
Logging in Kubernetes 1.19  has been migrated to the structured format. Two new methods are added to `k8s.io/klog/v2` library: `InfoS `and `ErrorS`. Should we update the scheduler plugins' logging calls to the structured logs and use the structured format in future logging? 

Refs
-  https://kubernetes.io/blog/2020/09/04/kubernetes-1-19-introducing-structured-logs/
-  https://github.com/kubernetes/klog

### PR DESCRIPTION
**Pull request by CIPHERTron**
### Files changed in `pkg/controller`:
```
pkg/controller/elasticquota.go:16
pkg/controller/podgroup.go:12
pkg/controller/elasticquota_test.go:2
```
This PR fixes #120

### CODE REVIEW

File: pkg/controller/elasticquota.go [hunk 1]
```diff
@@ -154,11 +154,11 @@ func (ctrl *ElasticQuotaController) processNextWorkItem() bool {
 	}
 	if err := ctrl.syncHandler(key); err != nil {
 		runtime.HandleError(err)
-		klog.Errorf("error syncing elastic quota %q: %s", key, err.Error())
+		klog.ErrorS(err, "error syncing elastic quota ", "elasticQuota", key)
```
Comment (Huang-Wei) [inline pkg/controller/elasticquota.go:157]:
```suggestion
		klog.ErrorS(err, "Error syncing elastic quota", "elasticQuota", key)
```

File: pkg/controller/elasticquota.go [hunk 2]
```diff
@@ -173,15 +173,15 @@ func (ctrl *ElasticQuotaController) syncHandler(key string) error {
 	// Get the elastic quota resource with this namespace/name
 	eq, err := ctrl.eqLister.ElasticQuotas(namespace).Get(name)
 	if apierrs.IsNotFound(err) {
-		klog.V(5).Infof("Elastic quota %q has been deleted ", key)
+		klog.V(5).InfoS("Elastic quota has been deleted ", "elasticQuota", key)
```
Comment (Huang-Wei) [inline pkg/controller/elasticquota.go:176]:
```suggestion
		klog.V(5).InfoS("ElasticQuota has been deleted", "elasticQuota", key)
```

File: pkg/controller/elasticquota.go [hunk 3]
```diff
@@ -173,15 +173,15 @@ func (ctrl *ElasticQuotaController) syncHandler(key string) error {
...
 		return nil
 	}
 	if err != nil {
-		klog.V(3).Infof("Unable to retrieve elastic quota %q from store: %v", key, err)
+		klog.V(3).ErrorS(err, "Unable to retrieve elastic quota from store", "elasticQuota", key)
 		return err
 	}
 
-	klog.V(5).Infof("Try to process elastic quota: %q", key)
+	klog.V(5).InfoS("Try to process elastic quota ", "elasticQuota", key)
```
Comment (Huang-Wei) [inline pkg/controller/elasticquota.go:184]:
```suggestion
	klog.V(5).InfoS("Try to process elastic quota", "elasticQuota", key)
```

File: pkg/controller/elasticquota.go [hunk 4]
```diff
@@ -228,7 +228,7 @@ func (ctrl *ElasticQuotaController) eqAdded(obj interface{}) {
 		runtime.HandleError(err)
 		return
 	}
-	klog.V(5).Infof("Enqueue new elastic quota key %q", key)
+	klog.V(5).InfoS("Enqueue new elastic quota key ", "elasticQuota", key)
```
Comment (Huang-Wei) [inline pkg/controller/elasticquota.go:231]:
```suggestion
	klog.V(5).InfoS("Enqueue new elastic quota key", "elasticQuota", key)
```

File: pkg/controller/elasticquota.go [hunk 5]
```diff
@@ -244,7 +244,7 @@ func (ctrl *ElasticQuotaController) eqDeleted(obj interface{}) {
 		runtime.HandleError(err)
 		return
 	}
-	klog.V(5).Infof("Enqueue deleted elastic quota key %q", key)
+	klog.V(5).InfoS("Enqueue deleted elastic quota key ", "elasticQuota", key)
```
Comment (Huang-Wei) [inline pkg/controller/elasticquota.go:247]:
```suggestion
	klog.V(5).InfoS("Enqueue deleted elastic quota key", "elasticQuota", key)
```

File: pkg/controller/podgroup.go [hunk 6]
```diff
@@ -93,14 +93,14 @@ func NewPodGroupController(client kubernetes.Interface,
 func (ctrl *PodGroupController) Run(workers int, stopCh <-chan struct{}) {
 	defer ctrl.pgQueue.ShutDown()
 
-	klog.Info("Starting Pod Group controller")
-	defer klog.Info("Shutting Pod Group controller")
+	klog.InfoS("Starting Pod Group controller", "podGroup")
+	defer klog.InfoS("Shutting Pod Group controller", "podGroup")
```
Comment (Huang-Wei) [inline pkg/controller/podgroup.go:97]:
```suggestion
	klog.InfoS("Starting Pod Group controller")
	defer klog.InfoS("Shutting Pod Group controller")
```

File: pkg/controller/podgroup.go [hunk 7]
```diff
@@ -93,14 +93,14 @@ func NewPodGroupController(client kubernetes.Interface,
...
 
 	if !cache.WaitForCacheSync(stopCh, ctrl.pgListerSynced, ctrl.podListerSynced) {
 		klog.Error("Cannot sync caches")
 		return
 	}
-	klog.Info("Pod Group sync finished")
+	klog.InfoS("Pod Group sync finished", "podGroup")
```
Comment (Huang-Wei) [inline pkg/controller/podgroup.go:103]:
```suggestion
	klog.InfoS("Pod Group sync finished")
```

File: pkg/controller/podgroup.go [hunk 8]
```diff
@@ -124,7 +124,7 @@ func (ctrl *PodGroupController) pgAdded(obj interface{}) {
 		pg.Status.ScheduleStartTime.Sub(pg.CreationTimestamp.Time) > 48*time.Hour {
 		return
 	}
-	klog.Infof("Enqueue key %v", key)
+	klog.InfoS("Enqueue key ", "enqueueKey", key)
```
Comment (Huang-Wei) [inline pkg/controller/podgroup.go:127]:
```suggestion
	klog.InfoS("Enqueue podGroup", "podGroup", key)
```

File: pkg/controller/podgroup.go [hunk 9]
```diff
@@ -142,10 +142,10 @@ func (ctrl *PodGroupController) podAdded(obj interface{}) {
 	}
 	pg, err := ctrl.pgLister.PodGroups(pod.Namespace).Get(pgName)
 	if err != nil {
-		klog.Error(err)
+		klog.ErrorS(err, "Error while adding pod")
 		return
 	}
-	klog.V(5).Infof("Add pg %v when pod %v add", pg.Name, pod.Name)
+	klog.V(5).InfoS("Add pod group when pod gets added", "podGroup", pg.Name, "pod", pod.Name)
```
Comment (Huang-Wei) [inline pkg/controller/podgroup.go:148]:
```suggestion
	klog.V(5).InfoS("Add pod group when pod gets added", "podGroup", klog.KObj(pg), "pod",  klog.KObj(pod))
```

File: pkg/controller/podgroup.go [hunk 10]
```diff
@@ -175,7 +175,7 @@ func (ctrl *PodGroupController) processNextWorkItem() bool {
 	}
 	if err := ctrl.syncHandler(key); err != nil {
 		runtime.HandleError(err)
-		klog.Errorf("error syncing pod group %q: %s", key, err.Error())
+		klog.ErrorS(err, "Error syncing pod group", "key", key)
```
Comment (Huang-Wei) [inline pkg/controller/podgroup.go:178]:
```suggestion
		klog.ErrorS(err, "Error syncing pod group", "podGroup", key)
```

File: pkg/controller/podgroup.go [hunk 11]
```diff
@@ -197,19 +197,19 @@ func (ctrl *PodGroupController) syncHandler(key string) error {
...
+		klog.V(3).ErrorS(err, "Unable to retrieve pod group from store", "podGroup", key)
 		return err
 	}
 
 	pgCopy := pg.DeepCopy()
 	selector := labels.Set(map[string]string{util.PodGroupLabel: pgCopy.Name}).AsSelector()
 	pods, err := ctrl.podLister.List(selector)
 	if err != nil {
-		klog.Errorf("List pods for group %v failed: %v", pgCopy.Name, err)
+		klog.ErrorS(err, "List pods for group failed", "podGroup", pgCopy.Name)
```
Comment (Huang-Wei) [inline pkg/controller/podgroup.go:212]:
```suggestion
		klog.ErrorS(err, "List pods for group failed", "podGroup", klog.KObj(pgCopy))
```

**Review by Huang-Wei — COMMENTED** [review]
Thanks for the PR. Some minor nits below.

**Comment by Huang-Wei** [comment]
Can you also link to the original issue # in the description?

**Comment by CIPHERTron** [comment]
@Huang-Wei I've made the requested changes. However, there's one more thing left which I wasn't able to figure it out.

File: pkg/controller/elasticquota.go [hunk 12]
```diff
@@ -111,20 +111,20 @@ func (ctrl *ElasticQuotaController) Run(workers int, stopCh <-chan struct{}) {
 	defer runtime.HandleCrash()
 	defer ctrl.eqQueue.ShutDown()
 	// Start the informer factories to begin populating the informer caches
-	klog.Info("Starting Elastic Quota control loop")
+	klog.InfoS("Starting Elastic Quota control loop")
 	// Wait for the caches to be synced before starting workers
-	klog.Info("Waiting for informer caches to sync")
+	klog.InfoS("Waiting for informer caches to sync")
 	if ok := cache.WaitForCacheSync(stopCh, ctrl.eqListerSynced, ctrl.podListerSynced); !ok {
 		klog.Fatalf("Cannot sync caches")
```
Comment (CIPHERTron) [inline pkg/controller/elasticquota.go:118]: @Huang-Wei I couldn't figure out how to migrate this to `ErrorS`. Could you please help me out here? Thanks!

Comment (Huang-Wei) [reply pkg/controller/elasticquota.go:118]: You can do sth like this:

```go
klog.ErrorS(nil, "Cannot sync caches")
os.Exit(1)
```

ref:L

- https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#replacing-fatal-calls
- https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#replacing-fatal-calls

Comment (CIPHERTron) [reply pkg/controller/elasticquota.go:118]: Ohh okay, I didn't know that we can pass nil in case of the `err` parameter. I'll make the necessary changes!

**Comment by CIPHERTron** [comment]
@Huang-Wei I've done all the required changes. I guess this PR is good to go now!

**Comment by Huang-Wei** [comment]
Thanks @CIPHERTron. Would you mind squashing the commits? I will approve after.

**Comment by CIPHERTron** [comment]
> Thanks @CIPHERTron. Would you mind squashing the commits? I will approve after.

This might be a dumb question probably but should I do an interactive rebase to squash commits because I've never squashed commits from my end 😅

**Comment by Huang-Wei** [comment]
> should I do an interactive rebase to squash commits

Yes, run `git rebase -i HEAD~3`. After that, the interactive window should show 3 commits. Then replace the 2nd and 3rd `pick` to `f` (short for `fixup`), and save. Lastly push the code using `-f` option.

The official doc can be found https://www.kubernetes.dev/docs/guide/github-workflow/.

**Comment by CIPHERTron** [comment]
> > should I do an interactive rebase to squash commits
> 
> Yes, run `git rebase -i HEAD~3`. After that, the interactive window should show 3 commits. Then replace the 2nd and 3rd `pick` to `f` (short for `fixup`), and save. Lastly push the code using `-f` option.
> 
> The official doc can be found https://www.kubernetes.dev/docs/guide/github-workflow/.

It's done @Huang-Wei. Thank you for such smooth experience. I got to learn a couple of new things through this PR 😸

**Pull request merged**